go/login/login.go

```diff
@@ -0,0 +1,388 @@
+)
+
+// Message struct represents the context required for authorization.
+// It contains HostIDType - type of identity, HostID - identity of the target (e.g. hostname, serial number, etc.),
+// Action - action that is being authorized.
+type Message struct {
+	HostIDType string
+	HostID     string
+	Action     string
+}
+
+// Construct returns a message from a Message according to the format: [<hostid-type>:]<hostid>[/<action>].
+// URL escaping is optional.
+func (m *Message) Construct(esc bool) []byte {
+	hostIDType := ""
+	hostID := ""
+	action := ""
+	if esc {
+		hostIDType = escape(m.HostIDType)
+		hostID = escape(m.HostID)
+	} else {
+		hostIDType = m.HostIDType
+		hostID = m.HostID
+	}
+
+	if hostIDType != "" {
+		hostIDType += ":"
+	}
+
+	if m.Action != "" {
+		action = "/" + m.Action
+	}
+	return []byte(hostIDType + hostID + action)
+}
+
+// Escapes the string so it can be safely placed inside a URL path segment,
+// replacing "/#?" special characters and not replacing "!*'();:@&=+$,[]" special characters.
+func escape(s string) string {
+	res := url.PathEscape(s)
+	for _, c := range "!*'();:@&=+$,[]" {
+		st := string(c)
+		strings.Replace(res, url.PathEscape(st), st, -1)
+	}
+	return res
+}
+
+// Handshake struct represents the context required for constructing the handshake.
+// It contains Prefix - either service key or its id, UserKey - user's public ephemeral key,
+// MessageTagPrefix - tag calculated under Message.
+type Handshake struct {
+	Prefix           byte
+	UserKey          glome.PublicKey
+	MessageTagPrefix []byte
+}
+
+// URLResponse struct represents the context required for the URL constructing.
+// It contains V - URL format version (currently always 1), HandshakeInfo - handshake info,
+// Msg - message info, d - glome.Dialog for the tag managing.
+type URLResponse struct {
+	V             byte
+	HandshakeInfo Handshake
+	Msg           Message
+	d             *glome.Dialog
+}
+
+// NewResponse returns a new URLResponse corresponding to the given arguments.
+func NewResponse(serviceKeyID uint8, serviceKey glome.PublicKey, userKey glome.PrivateKey,
+	V byte, hostIDType string, hostID string, action string, tagLen uint) (*URLResponse, error) {
+	var prefix byte
+	var r URLResponse
+
+	r.V = V
+
+	d, err := userKey.TruncatedExchange(&serviceKey, 1)
+	if err != nil {
+		return nil, err
+	}
+	r.d = d
+
+	r.Msg = Message{hostIDType, hostID, action}
+
+	if serviceKeyID == 0 {
+		// If no key ID was specified, send the first key byte as the ID.
+		prefix = serviceKey[0] & 0x7f
+	} else {
+		prefix = serviceKeyID & 0x7f
+	}
+	userPublic, err := userKey.Public()
+	if err != nil {
+		return nil, err
+	}
+	HandshakeInfo := Handshake{prefix, *userPublic, r.Tag(tagLen)}
+	r.HandshakeInfo = HandshakeInfo
+
+	return &r, nil
+}
+
+// ValidateAuthCode checks if the received tag corresponding to the base64-url encoded message constructed from the Message.
+// Returns true if the received tag is empty.
```

Comment:
I'm confused about the comment. What is it that the function is doing? When would you call it? Why is an empty tag true?

Reply:
Updated. The empty tag is ok for client's tag. Moved this logic to ParseURLResponse.